src/Protocol/ActivityPub.php

```diff
@@ -0,0 +1,1825 @@
+	 * @param integer $uid User ID
+	 * @param $trust_source
+	 */
+	private static function processActivity($activity, $body = '', $uid = null, $trust_source = false)
+	{
+		if (empty($activity['type'])) {
+			logger('Empty type', LOGGER_DEBUG);
+			return;
+		}
+
+		if (empty($activity['object'])) {
+			logger('Empty object', LOGGER_DEBUG);
+			return;
+		}
+
+		if (empty($activity['actor'])) {
+			logger('Empty actor', LOGGER_DEBUG);
+			return;
+
+		}
+
+		// $trust_source is called by reference and is set to true if the content was retrieved successfully
+		$object_data = self::prepareObjectData($activity, $uid, $trust_source);
+		if (empty($object_data)) {
+			logger('No object data found', LOGGER_DEBUG);
+			return;
+		}
+
+		if (!$trust_source) {
+			logger('No trust for activity type "' . $activity['type'] . '", so we quit now.', LOGGER_DEBUG);
+		}
+
+		switch ($activity['type']) {
+			case 'Create':
+			case 'Announce':
+				self::createItem($object_data, $body);
+				break;
+
+			case 'Like':
+				self::likeItem($object_data, $body);
+				break;
+
+			case 'Dislike':
+				self::dislikeItem($object_data, $body);
+				break;
+
+			case 'Update':
+				if (in_array($object_data['object_type'], ['Person', 'Organization', 'Service', 'Group', 'Application'])) {
+					self::updatePerson($object_data, $body);
+				}
+				break;
+
+			case 'Delete':
+				break;
+
+			case 'Follow':
+				self::followUser($object_data);
+				break;
+
+			case 'Accept':
+				if ($object_data['object_type'] == 'Follow') {
+					self::acceptFollowUser($object_data);
+				}
+				break;
+
+			case 'Undo':
+				if ($object_data['object_type'] == 'Follow') {
+					self::undoFollowUser($object_data);
+				} elseif (in_array($object_data['object_type'], ['Like', 'Dislike', 'Accept', 'Reject', 'TentativeAccept'])) {
+					self::undoActivity($object_data);
+				}
+				break;
+
+			default:
+				logger('Unknown activity: ' . $activity['type'], LOGGER_DEBUG);
+				break;
+		}
+	}
+
+	/**
+	 * @brief 
+	 *
+	 * @param array $activity
+	 * @param $actor
+	 *
+	 * @return 
+	 */
+	private static function getReceivers($activity, $actor)
+	{
+		$receivers = [];
+
+		// When it is an answer, we inherite the receivers from the parent
+		$replyto = JsonLD::fetchElement($activity, 'inReplyTo', 'id');
+		if (!empty($replyto)) {
+			$parents = Item::select(['uid'], ['uri' => $replyto]);
+			while ($parent = Item::fetch($parents)) {
+				$receivers['uid:' . $parent['uid']] = $parent['uid'];
+			}
+		}
+
+		if (!empty($actor)) {
+			$profile = APContact::getProfileByURL($actor);
+			$followers = defaults($profile, 'followers', '');
+
+			logger('Actor: ' . $actor . ' - Followers: ' . $followers, LOGGER_DEBUG);
+		} else {
+			logger('Empty actor', LOGGER_DEBUG);
+			$followers = '';
+		}
+
+		$elements = ['to', 'cc', 'bto', 'bcc'];
+		foreach ($elements as $element) {
+			if (empty($activity[$element])) {
+				continue;
+			}
+
+			// The receiver can be an arror or a string
```

Comment:
`array`

Reply:
Maybe "arror" is some "error" in an "array"? 😬